helix-view/src/document.rs

```diff
@@ -186,47 +187,47 @@ impl Document {
         Ok(doc)
     }
 
-    // TODO: remove view_id dependency here
-    pub fn format(&mut self, view_id: ViewId) {
-        if let Some(language_server) = self.language_server() {
-            // TODO: await, no blocking
-            let transaction = helix_lsp::block_on(
+    pub fn format(&self) -> BoxFuture<'static, anyhow::Result<Transaction>> {
+        let text = self.text().clone();
+        let language_server = self.language_server.clone();
+        let identifier = self.identifier().clone();
+        if let Some(language_server) = language_server {
+            async move {
                 language_server
-                    .text_document_formatting(self.identifier(), lsp::FormattingOptions::default()),
-            )
-            .map(|edits| {
-                helix_lsp::util::generate_transaction_from_edits(
-                    self.text(),
-                    edits,
-                    language_server.offset_encoding(),
-                )
-            });
-
-            if let Ok(transaction) = transaction {
-                self.apply(&transaction, view_id);
-                self.append_changes_to_history(view_id);
+                    .text_document_formatting(identifier, lsp::FormattingOptions::default())
+                    .await
+                    .map(|edits| {
```

Comment:
A cheap trick would be to send through the `text` + `edits` and generate a transaction outside the future, thus avoiding the `Send`/`Sync` requirement.

Reply:
Thanks! I'll give that a try.